File: algorithms/275. H-Index II/solution.py
class Solution(object):
    def hIndex(self, citations):
        """
        :type citations: List[int]
        :rtype: int
        """
        if not citations:
            return 0
        n = len(citations)
        l = 0; r = n - 1
        while l < r:
            m = (l + r) / 2
            if citations[m] >= n - m:
                r = m
            else:
                l = m + 1
        if citations[l] >= n - l:
            return n - l
        return 0

File: algorithms/275. H-Index II/solution.cc
class Solution {
public:
    int hIndex(vector<int>& citations) {
        int n = citations.size(), l = 0, r = n - 1, m;
        while ( l <= r ) {
            m = l + (r - l) / 2;
            if ( citations[m] > n - m ) {
                r = m - 1;
            } else if (citations[m] < n - m ) {
                l = m + 1;
            } else {
                return n - m;
            }
        }
        return n - r - 1;
    }
};
